File: content-solutions/implementation/sim-zip-best.js
function solution(s) {
  const pieceLen = (chunk, cnt) => chunk.length + (cnt > 1 ? String(cnt).length : 0);
  let best = s.length;
  for (let k = 1; k <= s.length; k++) {
    const chunks = [];
    for (let i = 0; i < s.length; i += k) chunks.push(s.slice(i, i + k));
    let length = 0;
    let prev = chunks[0];
    let cnt = 1;
    for (let i = 1; i < chunks.length; i++) {
      if (chunks[i] === prev) cnt++;
      else {
        length += pieceLen(prev, cnt);
        prev = chunks[i];
        cnt = 1;
      }
    }
    length += pieceLen(prev, cnt);
    best = Math.min(best, length);
  }
  return best;
}

File: content-solutions/implementation/sim-zip-best.java
import java.util.*;

class Solution {
    int pieceLen(String chunk, int cnt) {
        return chunk.length() + (cnt > 1 ? String.valueOf(cnt).length() : 0);
    }

    public int solution(String s) {
        int best = s.length();
        for (int k = 1; k <= s.length(); k++) {
            List<String> chunks = new ArrayList<>();
            for (int i = 0; i < s.length(); i += k) chunks.add(s.substring(i, Math.min(s.length(), i + k)));
            int length = 0, cnt = 1;
            String prev = chunks.get(0);
            for (int i = 1; i < chunks.size(); i++) {
                if (chunks.get(i).equals(prev)) cnt++;
                else {
                    length += pieceLen(prev, cnt);
                    prev = chunks.get(i);
                    cnt = 1;
                }
            }
            length += pieceLen(prev, cnt);
            best = Math.min(best, length);
        }
        return best;
    }
}
